client/src/receiver.rs

```diff
@@ -0,0 +1,452 @@
+
+use crate::{error::DenimProtocolError, message::create_message};
+
+pub struct DenimReceiver<T: SendingBuffer> {
+    client: Arc<Mutex<WebSocketClient>>,
+    enqueue_sam_envelope: Sender<ServerEnvelope>,
+    enqueue_sam_status: Sender<ServerStatus>,
+    enqueue_chunks: Option<Sender<DenimChunk>>,
+    sending_buffer: T,
+}
+
+impl<T: SendingBuffer> DenimReceiver<T> {
+    pub fn new(
+        client: Arc<Mutex<WebSocketClient>>,
+        enqueue_sam_envelope: Sender<ServerEnvelope>,
+        enqueue_sam_status: Sender<ServerStatus>,
+        sending_buffer: T,
+    ) -> Self {
+        Self {
+            client,
+            enqueue_sam_envelope,
+            enqueue_sam_status,
+            enqueue_chunks: None,
+            sending_buffer,
+        }
+    }
+
+    async fn send_ack(&mut self, id: MessageId) -> Result<(), DenimProtocolError> {
+        let msg = create_message(
+            &mut self.sending_buffer,
+            ClientMessage::builder()
+                .id(id.into())
+                .r#type(ClientMessageType::ClientAck.into())
+                .build(),
+        )
+        .await?;
+        self.client
+            .lock()
+            .await
+            .send(Message::Binary(msg.encode_to_vec().into()))
+            .await
+            .map_err(DenimProtocolError::from)
+    }
+
+    async fn handle_sam_message(
+        &mut self,
+        message: ServerMessage,
+    ) -> Result<(), DenimProtocolError> {
+        let res = match EnvelopeOrStatus::try_from(message)? {
+            EnvelopeOrStatus::Envelope(id, envelope) => self.dispatch_envelope(id, envelope).await,
+            EnvelopeOrStatus::Status(status) => self.dispatch_server_status(status).await,
+        };
+
+        match res {
+            Ok(Some(id)) => self.send_ack(id).await,
+            Ok(None) => Ok(()),
+            Err(e) => Err(e),
+        }
+    }
+
+    async fn dispatch_envelope(
+        &mut self,
+        id: MessageId,
+        envelope: ServerEnvelope,
+    ) -> Result<Option<MessageId>, DenimProtocolError> {
+        self.enqueue_sam_envelope
+            .send(envelope)
+            .await
+            .inspect_err(|e| debug!("{e}"))
+            .map_err(|_| DenimProtocolError::WebSocketError(WebSocketError::Disconnected))
+            .map(|_| Some(id))
+    }
+
+    async fn dispatch_server_status(
+        &mut self,
+        status: ServerStatus,
+    ) -> Result<Option<MessageId>, DenimProtocolError> {
+        self.enqueue_sam_status
+            .send(status)
+            .await
+            .inspect_err(|e| debug!("{e}"))
+            .map_err(|_| DenimProtocolError::WebSocketError(WebSocketError::Disconnected))
+            .map(|_| None)
+    }
+
+    async fn handle_chunks(&mut self, chunks: Vec<DenimChunk>) {
+        if let Some(sender) = &self.enqueue_chunks {
+            for chunk in chunks {
+                match sender.send(chunk).await {
+                    Ok(_) => continue,
```

Comment:
Should put the chunks into the receiving buffer proces_chunk(), and only send the denim message when it is fully built

Reply:
this will be fixed in #25